src/framework/codemirror-lang-typeql/typeQLAutocompleteSchema.ts

```diff
@@ -0,0 +1,246 @@
+        tree.iterate({
+            enter: (cursor: TreeCursor) => {
+                let node = cursor.node;
+                if (node.type?.id === tokens.DefinitionType) {
+                    if (node.firstChild?.type?.id === tokens.KIND) {
+                        let labelNode = node.firstChild!.nextSibling!;
+                        let label = extractText(text, labelNode.from, labelNode.to);
+                        let kind = extractText(text, node.firstChild!.from, node.firstChild!.to);
+                        switch (kind) {
+                            case "entity": {
+                                builder.objectType(label);
+                                 break;
+                            }
+                            case "relation": {
+                                builder.objectType(label);
+                                 break;
+                            }
+                            case "attribute": {
+                                builder.attributeType(label);
+                                 break;
+                            }
+                        }
+                    }
+                }
+            }
+        });
+
+        // Extract owns/relates/plays. Idk what to do with sub or annotations.
+        tree.iterate({
+            enter: (cursor: TreeCursor) => {
+                let node = cursor.node;
+                if (node.type.id === tokens.DefinitionType) {
+                    let labelNode = (node.firstChild?.type?.id === tokens.KIND) ? node.firstChild!.nextSibling! : node.firstChild!;
+                    let label = extractText(text, labelNode.from, labelNode.to);
+                    node.getChildren(tokens.TypeCapability).forEach((typeCapabilityBaseNode: SyntaxNode) => {
+                        let actualCapabilityNode = typeCapabilityBaseNode.firstChild!.firstChild!;
+                        switch (actualCapabilityNode.type.id) {
+                            // We actually only want type-declarations for now.
+                            case tokens.RelatesDeclaration: {
+                                let roleTypeNode = actualCapabilityNode.firstChild!.nextSibling!;
+                                let roleType = extractText(text, roleTypeNode.from, roleTypeNode.to);
+                                builder.recordRelates(label, `${label}:${roleType}`);
+                                break;
+                            }
+                            default: {
+                                // Ignore other capabilities for now
+                                break;
+                            }
+                        }
+                    });
+                }
+            }
+        });
+        return builder.build();
+    }
+
+    static fromDriver(driver: DriverState, database: string): TypeQLAutocompleteSchemaImpl | null {
+        function runQuery(driver: DriverState, database: string, query: string): ConceptRowAnswer[] {
+            // todo: Help please alex
```

Comment:
@alexjpwalker , This is the function I don't know how to implement in the current state of studio. Can you sketch me an implementation? (or do it if you think it's faster that way).